### PR TITLE
Level DB support for Whisper

### PR DESCRIPTION
**Pull request by gluk256**

### CODE REVIEW

**Comment by gluk256** [comment]
never mind the errors -- they are not mine

File: libwhisper/WhisperDB.cpp [hunk 1]
```diff
@@ -134,3 +135,54 @@ void WhisperDB::loadAll(std::map<h256, Envelope>& o_dst)
...
+		res.appendRaw(meta.out());
+		res.appendRaw(msg.out());
+
+		bytes b;
+		res.swapOut(b);
+		insert(_key, b);
+	}
+	catch(RLPException const& ex)
+	{
+		cwarn << "RLPException in WhisperDB::save():" << ex.what();
```
Comment (arkpar) [inline libwhisper/WhisperDB.cpp]: For boost exceptions diagnostic_information or current_exception_diagnostic_information should be used.

File: libwhisper/WhisperDB.cpp [hunk 2]
```diff
@@ -134,3 +135,54 @@ void WhisperDB::loadAll(std::map<h256, Envelope>& o_dst)
...
+{
+	try
+	{
+		RLPStream meta;
+		streamMetaInfo(meta, _e);
+
+		RLPStream msg;
+		_e.streamRLP(msg);
+
+		RLPStream res(2);
```
Comment (arkpar) [inline libwhisper/WhisperDB.cpp]: use rlpList(...) instead

File: libwhisper/WhisperHost.cpp [hunk 3]
```diff
@@ -69,13 +69,15 @@ void WhisperHost::inject(Envelope const& _m, WhisperPeer* _p)
...
-		m_expiryQueue.insert(make_pair(_m.expiry(), h));
+		if (!_m.isStoreForever())
+			m_expiryQueue.insert(make_pair(_m.expiry(), h));
 	}
 
 	// rating of incoming message from remote host is assessed according to the following criteria:
 	// 1. installed watch match; 2. bloom filter match; 2. ttl; 3. proof of work
 
 	int rating = 0;
+	bool match_watch = false;
```
Comment (arkpar) [inline libwhisper/WhisperHost.cpp]: camelCase

File: libwhisper/Message.h [hunk 4]
```diff
@@ -75,12 +76,12 @@ class Envelope
...
 	bytes const& data() const { return m_data; }
 
-	Message open(Topics const& _t, Secret const& _s = Secret()) const;
-
-	unsigned workProved() const;
-	void proveWork(unsigned _ms);
-
 	bool matchesBloomFilter(TopicBloomFilterHash const& f) const;
+	void setStoreForever() { m_storeForever = true; }
+	bool isStoreForever() const { return m_storeForever; }
```
Comment (arkpar) [inline libwhisper/Message.h]: isStoredForeved or storeForever

**Pull request merged**